Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php

```diff
@@ -218,22 +221,35 @@ protected function requireNodeTypeToAllowNodesOfTypeInReference(
         NodeTypeName $nodeTypeNameInQuestion
     ): void {
         $nodeType = $this->requireNodeType($nodeTypeName);
-        $propertyDeclaration = $nodeType->getProperties()[$referenceName->value] ?? null;
-        if (is_null($propertyDeclaration)) {
-            throw ReferenceCannotBeSet::becauseTheNodeTypeDoesNotDeclareIt($referenceName, $nodeTypeName);
-        }
-
-        $constraints = $propertyDeclaration['constraints']['nodeTypes'] ?? [];
+        $constraints = $nodeType->getReferences()[$referenceName->value]['constraints']['nodeTypes'] ?? [];
 
         if (!ConstraintCheck::create($constraints)->isNodeTypeAllowed($this->requireNodeType($nodeTypeNameInQuestion))) {
-            throw ReferenceCannotBeSet::becauseTheConstraintsAreNotMatched(
+            throw ReferenceCannotBeSet::becauseTheNodeTypeConstraintsAreNotMatched(
                 $referenceName,
                 $nodeTypeName,
                 $nodeTypeNameInQuestion
             );
         }
     }
 
+    protected function requireNodeTypeToAllowCountOfReferencesInReference(SerializedNodeReferences $nodeReferences, ReferenceName $referenceName, NodeTypeName $nodeTypeName): void
```

Comment:
I'd suggest NumberOf instead of CountOf, since we handle references, not votes in an election

Reply:
if only this was my own PR :D